app/src/main/java/dev/msfjarvis/lobsters/ui/viewmodel/LobstersViewModel.kt

```diff
@@ -4,49 +4,62 @@ import androidx.lifecycle.ViewModel
 import androidx.lifecycle.viewModelScope
 import androidx.paging.Pager
 import androidx.paging.PagingConfig
+import androidx.paging.cachedIn
 import dagger.hilt.android.lifecycle.HiltViewModel
+import dev.msfjarvis.lobsters.data.api.LobstersApi
+import dev.msfjarvis.lobsters.data.local.LobstersPost
 import dev.msfjarvis.lobsters.data.remote.LobstersPagingSource
-import dev.msfjarvis.lobsters.data.source.PostsDatabase
-import dev.msfjarvis.lobsters.model.LobstersPost
+import dev.msfjarvis.lobsters.data.repo.LobstersRepository
 import kotlinx.coroutines.flow.MutableStateFlow
-import kotlinx.coroutines.flow.StateFlow
-import kotlinx.coroutines.flow.collectLatest
+import kotlinx.coroutines.flow.asStateFlow
 import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 @HiltViewModel
 class LobstersViewModel @Inject constructor(
   private val pagingSource: LobstersPagingSource,
-  database: PostsDatabase,
+  private val lobstersRepository: LobstersRepository,
+  private val lobstersApi: LobstersApi,
 ) : ViewModel() {
   private val _savedPosts = MutableStateFlow<List<LobstersPost>>(emptyList())
-  private val savedPostsDao = database.savedPostsDao()
-  val savedPosts: StateFlow<List<LobstersPost>> get() = _savedPosts
+  private lateinit var source: LobstersPagingSource
+  val savedPosts = _savedPosts.asStateFlow()
   val posts = Pager(PagingConfig(25)) {
-    pagingSource
-  }.flow
+    source = LobstersPagingSource(lobstersApi, lobstersRepository)
+    source
+  }.flow.cachedIn(viewModelScope)
 
-  init {
-    getSavedPosts()
+  fun toggleSave(post: LobstersPost) {
+    viewModelScope.launch {
+      val isSaved = lobstersRepository.isPostSaved(post.short_id)
+      if (isSaved) removeSavedPost(post) else savePost(post)
+      invalidateSource()
```

Comment:
I'm assuming this invalidation is what is causing the entire list to be redrawn?

Reply:
Yep, I have a few ideas in mind to fix this :thinking: